### DISCUSSION TITLE
Does pyUSB handle packet fragmentation in bulk write/read methods?

### DISCUSSION THREAD

**Discussion by brianreinhold** [pyusb/pyusb#427] [Q&A] [answered]
I could not find it stated anywhere, but in the bulk in/out write and read methods all that is stated is that one passes in a data array. Does pyUSB handle the fragmentation or is the application responsible for doing it?

In other words, do I need to take an array of arbitrary size and break it into (usually 64) byte packets or can I send the entire packet of length M and pyUSB handles the fragmentation?

**Comment by jonasmalacofilho** [comment]
_That's a great question, and I'm surprised it doesn't come up more often.  I also didn't quite remember how it worked for all transfer types.  I guess most of us are working with simple/interrupt protocols and/or only specific devices.  Anyway..._

The caller is responsible for chucking or calling `read()`/`write()` as many time as necessary.  You don't necessarily need to precompute the chunks, you can also call `read()`/`write()` in a loop, checking the return value and adjusting the parameters accordingly.

In this (and other regards) PyUSB works quite similarly to the corresponding LibUSB (1.0) functions.

**Reply by brianreinhold** [reply]
Okay, so I see that the max transfer size is 64 bytes so I will need to break up my array (up to 2046 bytes) into 64 byte hunks and if the last hunk happens to be 64 bytes I need to send a final time with 0 bytes. Same with the read ( I need to read until I get a 0-byte read or a read of less than 64 bytes). This is the correct behavior ...

**Reply by jonasmalacofilho** [reply]
You know, 64 bytes seems quite small for a bulk transfer... for chunks that small I would generally expect the designer to have picked interrupt transfers (even though there are other reasons to pick one or the other).

I also don't understand why you need a final 0-sized transfer... can't you simply not send it?

**Reply by brianreinhold** [reply]
At low levels a bulk transfer is usually done in 64-byte hunks. The actual max size of an individual transfer is given by the device' descriptors. The transfer is done when a packet size of less than 64 bytes is received or a 0-byte packet is received. The 0-byte transfer is needed in case the actual byte array size is a multiple of 64. When I was using libusb some time ago (2008) in C I had to handle the chunking and 0-byte case for both the bulk read and writes at the application level. That is what I am asking about here.

This link might help: [bulk transfers](https://www.beyondlogic.org/usbnutshell/usb4.shtml#Bulk)

Below are the descriptors for the device I am working with. You will see the max transfer size in each of the endpoint descriptors (this device only supports bulk transfers) is 64 bytes:

Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.01
  bDeviceClass            0
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0        64
  idVendor           0x18d1 Google Inc.
  idProduct          0x9102
  bcdDevice            2.00
  iManufacturer           1 Verily
  iProduct                2 Study Watch
  iSerial                 3 C2QT510X2052004F
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0020
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              300mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      1
      bInterfaceProtocol    255
      iInterface              0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x01  EP 1 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        **wMaxPacketSize     0x0040  1x 64 bytes**
        bInterval               1
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        **wMaxPacketSize     0x0040  1x 64 bytes**
        bInterval               1

**Reply by brianreinhold** [reply]
Any answer to this? It turns out that the usb1 package handles the fragmentation. It's probably done here as well, but I am not sure. I don't see any examples for bulk transfers of large packets (or at least greater than the max low level packet chunk).

**Reply by jonasmalacofilho** [reply]
> Any answer to this?

Can you clarify what part of your previous comment you want me to address?

> It turns out that the usb1 package handles the fragmentation.

I think there might be an issue with what "handle fragmentation" is being understood for.

Like I said previously, you don't need to preemptively calculate the appropriate chuck size, _as long as you call `read()`/`write()` as many times as necessary (i.e. until all bytes have been read/written).  And this appears to be the same with PyUSB and `usb1`.

Now, hypothetically, the underlying libraries or the kernel could have their APIs automatically do as many underlying requests as necessary.  But of the top of my head I don't remember having seen that.

> It's probably done here as well, but I am not sure.

Like I said above: 

The caller is responsible for chucking or calling `read()`/`write()` as many time as necessary. You don't necessarily need to precompute the chunks, _you can also call `read()`/`write()` in a loop, checking the return value and adjusting the parameters accordingly._

P.S. The 64 byte maximum for bulk packets is quite old, newer USB revisions allow up to 1024 bytes.  But like I said, there are valid reasons to use bulk transfers even with small packet sizes, I'm just not used to seeing them in practice.

**Reply by brianreinhold** [reply]
The device I have has megabytes to transfer. But it has a bulk transfer size of 64. All I wanted to know is if I need to do a loop over 64 byte hunks like at low levels or do I just pass in the array of bytes. It sounds like the latter.

Yes, USB 2.0 supports 512 and 3.0 supports 1024. These sizes are still much smaller than the amount of data I need to transfer. However, the device does fine with the 64 byte bulk transfer limit.

SO I will conclude that I do not need to loop over fragments of size 64 when I invoke the write method.

**Reply by jonasmalacofilho** [reply]
Actually, I was wrong in one import point: both the Linux Kernel documentation and LibUSB's implementation for Linux mention automatic splitting of large buffers into smaller underlying packets.

For that reason, you really shouldn't pre-compute the chunks yourself, as they might even do it more efficiently (for example, in some scenarios LibUSB may submit multiple URBs at once).

That said, both API specs (and PyUSB) might still return early with actual < requested transferred bytes.

**Reply by brianreinhold** [reply]
If the underlying software is doing it, I have no intention to do it. I just wanted to be sure that I didn't have to do it. And yes, I will check the returned 'bytes_read' value!

**Comment by brianreinhold** [comment, marked as answer]
The answer is that it does - so you do not have to do it!
